libazureinit/src/goalstate.rs

```diff
@@ -41,34 +45,72 @@ pub struct RoleInstance {
     instance_id: String,
 }
 
-pub async fn get_goalstate(client: &Client) -> Result<Goalstate, Error> {
-    let url = "http://168.63.129.16/machine/?comp=goalstate";
+const DEFAULT_GOALSTATE_URL: &str =
+    "http://168.63.129.16/machine/?comp=goalstate";
+
+pub async fn get_goalstate(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    url: Option<&str>,
+) -> Result<Goalstate, Error> {
+    let url = url.unwrap_or(DEFAULT_GOALSTATE_URL);
 
     let mut headers = HeaderMap::new();
     headers.insert("x-ms-agent-name", HeaderValue::from_static("azure-init"));
     headers.insert("x-ms-version", HeaderValue::from_static("2012-11-30"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            let mut rest_timeout = total_timeout;
 
-    if response.status().is_success() {
-        let body = response.text().await?;
+            if let Ok(response) = client
+                .get(url)
+                .headers(headers.clone())
+                .timeout(Duration::from_secs(http::WIRESERVER_HTTP_TIMEOUT_SEC))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
 
-        let goalstate: Goalstate = from_str(&body)?;
-        Ok(goalstate)
-    } else {
-        Err(Error::HttpStatus {
-            endpoint: url.to_owned(),
-            status: response.status(),
-        })
-    }
+                if statuscode.is_success() && statuscode == StatusCode::OK {
```

Comment:
this block of code is repeated three times between goalstate, imds, and report health. I would recommend refactoring this into http module so that we can pass in arguments like url, total_timeout, per call timeout, retry codes, etc...
I'm ok with the PR going in the way it is and the refactor to reduce redundant code happening in a follow-up PR

Reply:
Agreed in general.
Looking at the code once again, `.is_success()` is actually not needed, because `success` means code from 200 to 300.
So I simply removed the first condition, and kept only the second.